### ISSUE TITLE
Cannot walk a list of files and pass them to a subprocess (if any are non-UTF8)

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#33368]
It is currently not possible to walk a list of files and, for each one, invoke a subprocess, passing it the file name. Specifically, it's not possible to do this for non-UTF-8 file names. Such file names are common on non-English systems and are easily created by malicious actors.

Similarly, it's not possible to write Dart-based shell scripts for manipulating symlinks, etc, in environments that are potentially hostile (or legacy environments that potentially have non-UTF-8 strings).

To resolve this, several APIs need to have variants that work with byte arrays instead of strings, including:

 - `FileSystemEntity.path`.
 - `Process.run`, `Process.runSync`, and `Process.start`, specifically the `executable`, `workingDirectory`, `arguments`, and `environment` arguments.
 - `FileSystemEntity.rename` and `FileSystemEntity.renameSync`
 - `FileSystemEntity.resolveSymbolicLinks` and `FileSystemEntity.resolveSymbolicLinksSync`
 - All the static methods on `FileSystemEntity`.
 - All the new methods on `Link`.
 - Some of the `Directory` and `File` APIs.

It may be worth considering a language-level change wherein the existing String features can handle non-UTF-8 natively rather than having to duplicate the entire String-based API surface.

cc @leafpetersen, @bkonyi

**Comment by leafpetersen** [comment]
> Specifically, it's not possible to do this for non-UTF-8 file names.

@Hixie Can you clarify what you mean?  What is a non-UTF-8 file name?  Presumably all file names can be encoded in UTF-8.

**Comment by Hixie** [comment]
File names are byte sequences. They are not strings. There is no guarantee that they will be UTF-8.

For example, this command creates an empty file with a non-UTF-8 file name:

```
touch `echo -e '\x80\x7f.txt'`
```

**Comment by leafpetersen** [comment]
Got it thanks.  That's quite unpleasant.  It would be nice to have the string abstraction represent, well, strings.  

My next question would be, is this functionality that we need to support with the same API?  Do we really want to/need to be able to use all of these standard APIs to manipulate non-string files?   Is this something that ordinary users need to be able to do?  Are we not potentially causing more trouble by exposing them to things that we call "Strings" which are not valid strings?

I guess the problematic cases are things like listing the contents of a directory which contains a non-string file name.  What does `dart:io` do with that now?

cc @zanderso @lrhn @nex3 @rakudrama

**Comment by Hixie** [comment]
Ideally we'd have some sort of type that was as convenient as strings (had a literal format, e.g. `b'foo.txt'`, had a `toString()` that defaulted to a fault-tolerant conversion to UTF-8, etc) but was just a `Uint8List` or `ByteData` or some such. Then that would be the type that we would want to use for all APIs that dealt with filesystem paths.

**Comment by nex3** [comment]
I have no strong opinion on this beyond a general desire to keep the filesystem APIs as usable as they are today.

**Comment by zanderso** [comment]
Listing a directory returns a stream or list of `FileSystemEntitie`s, which internally store the path as a `Uint8List`. When you say e.g. `File.path`, we convert that to a String using `utf8.decode(..., allowMalformed: true)`. There's currently no way to extract the `Uint8List` raw path as adding a new getter to `FileSystemEntity` would break implementers in e.g. `package:file`.

We could add a new type to `dart:io` that behaves as @Hixie describes, but I think the APIs would become less usable without the ability to accept string-like literals.

**Comment by Hixie** [comment]
Hence my suggestion of a new string-like literal for `Uint8List`. :-)

**Comment by leafpetersen** [comment]
@zanderso Just to be sure I understand the implications for dart:io, could you expand a little on what combination of language features + breaking changes would be required to make this work well?

**Comment by zanderso** [comment]
@leafpetersen sorry for the delayed response. I'm not sure I can add very much new to what @Hixie has already described. But taking a closer look at the example in the issue title, we could look at `Process.run`. Today, the path to the binary and the arguments are `String`s:
```dart
Future<ProcessResult> run (
    String executable,
    List<String> arguments, {
    ...
})
```
So the problem in this issue is that you can't take a non-utf8 path and pass it to a program because the arguments are `List<String>`

A common pattern is also to do:
```dart
var result = await Process.run("path/to/some/program", [
  "--arg1",
  "--arg2",
]);
```

The breaking API change would be to change the signature of `Process.run` to something like:
```dart
Future<ProcessResult> run (
   Bytes executable,
    List<Bytes> arguments, {
    ...
})
```

Where `Bytes` is really just a `Uint8List`. But then to do the common pattern, you'd have to write:
```dart
var result = await Process.run(Bytes.fromString("path/to/some/program"), [
  Bytes.fromString("--arg1"),
  Bytes.fromString("--arg2"),
]);
```

So, instead of writing `Bytes.fromString` everywhere, it would be nice to be able to write a literal for `Bytes`, like b'this/is/a/path'.

Alternately, `String` could be a subtype of `Bytes` restricted to representing utf8 strings. Then even though `Process.run` takes `Bytes`, you could still pass it `String` literals.

You might also just say, why not keep the old `String`-based call and add a new call:
```dart
Process.runWithBytes(Bytes executable, List<Bytes> arguments)
```
for the rare case where something is non-utf8. Then you get into situations like:
```dart
await for (FileSystemEntity fse in someDirectory.list()) {
  if (hasUtf8Path(fse)) {
    Process.run(..., /*String*/ fse.path, ...);
  } else {
    Process.runWithBytes(..., /*Bytes*/ fse.rawPath, ...);
    // But also requires converting everything in the ... to Bytes as well.
  }
}
```
instead of just:
```dart
await for (FileSystemEntity fse in someDirectory.list()) {
  Process.run(..., fse.path, ...);
}
```

Anyway, to summarize, to work well, I think the signatures of `dart:io` functions would have to change from accepting `String` to accepting `Bytes`, and either there is a literal for `Bytes`, or `String` is a subtype of `Bytes`.

**Comment by lrhn** [comment]
It seems to me that the byte-list-to-string conversion should be parameterized with an `Encoding`. What hapens here is that we default to UTF-8 without having a way to choose a non-default decoder, and not all byte lists are actually UTF-8 encodings.
If you could ask for `fse.pathString(latin1)` instead, you would be able to get all the bytes interpreted the way you want them.

That would also require an `Encoding` argument to all functions that *accept* String arguments, since strings and byte lists are separate concepts, and there is more than one way to convert between the two.
Using UTF-8 as default is fine, it's just not always sufficient.

So `File.rename` would be `Future<File> rename(String newPath, {Encoding encoding});`

Maybe it would be safer to use `allowInvalid: false` and throw when the bytes clearly aren't UTF-8. Then the user can catch that and try something else (Latin-1 always "works", it might not print the way the creator of the file expected).

**Comment by Hixie** [comment]
File systems are not technically encoded strings, they are raw bytes. Strings are not byte sequences, they are arrays of Unicode scalar values.

When you read a filename from the system, you may not have any way to guess his to decode it; if you want to round trip such a filename, you should not need to pick an arbitrary encoding to convert the bytes from and to.

When you have a String, it should be possible to have constructed it from multiple differently-encoded buffers, such that there is no single encoding that represents it.

**Comment by lrhn** [comment]
Marking this as a language enhancement request.
Could also be a library enhancement request or a library bug, depending on perspective.

**Comment by lrhn** [comment]
Not really a language feature unless we plan to change `String` to allow raw bytes. (We don't.)
So it's a `dart:io` API issue that you cannot specify either file names as bytes, or an encoding for file names (where, if you choose `latin1`, you can treat the strings as byte sequences).
